File: api/src/main/java/com/example/movie/api/config/GlobalExceptionHandler.java
package com.example.movie.api.config;

import static org.springframework.http.HttpStatus.BAD_REQUEST;
import static org.springframework.http.HttpStatus.INTERNAL_SERVER_ERROR;
import static org.springframework.http.HttpStatus.UNSUPPORTED_MEDIA_TYPE;

import com.example.movie.common.api.ApiResponse;
import jakarta.validation.ConstraintViolationException;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.http.converter.HttpMessageNotReadableException;
import org.springframework.web.HttpMediaTypeNotSupportedException;
import org.springframework.web.HttpRequestMethodNotSupportedException;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.MissingServletRequestParameterException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;
import org.springframework.web.method.annotation.MethodArgumentTypeMismatchException;

@Slf4j
@RestControllerAdvice
public class GlobalExceptionHandler {

    @ExceptionHandler({
        HttpMediaTypeNotSupportedException.class,
        HttpRequestMethodNotSupportedException.class
    })
    public ResponseEntity<ApiResponse<?>> unSupportedMediaTypeException() {
        HttpStatus responseType = UNSUPPORTED_MEDIA_TYPE;

        return ResponseEntity
            .status(responseType)
            .body(ApiResponse.exception(responseType));
    }

    /**
     * 잘못된 API 요청 Exception
     */
    @ExceptionHandler({
        ConstraintViolationException.class,
        MethodArgumentNotValidException.class,
        MethodArgumentTypeMismatchException.class,
        HttpMessageNotReadableException.class,
        MissingServletRequestParameterException.class,
    })
    public ResponseEntity<ApiResponse<?>> badRequestException(Exception e) {
        log.error(e.getMessage(), e);

        HttpStatus responseType = BAD_REQUEST;

        return ResponseEntity
            .status(responseType)
            .body(ApiResponse.exception(responseType));
    }

    /**
     * 예기치 않은 서버 예외
     */
    @ExceptionHandler({
        Exception.class,
        IllegalStateException.class
    })
    public ResponseEntity<ApiResponse<?>> unknownException(Exception e) {
        log.error(e.getMessage(), e);
        HttpStatus responseType = INTERNAL_SERVER_ERROR;

        return ResponseEntity
            .status(responseType)
            .body(ApiResponse.exception(responseType));
    }
}

File: module-common/src/main/kotlin/com/example/redis/exceptions/GlobalExceptionHandler.kt
package com.example.redis.exceptions

import com.example.redis.response.ApiResponse
import com.example.redis.response.ErrorResponse
import jakarta.validation.ConstraintViolationException
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.validation.FieldError
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice

@RestControllerAdvice
class GlobalExceptionHandler {
    @ExceptionHandler(MethodArgumentNotValidException::class)
    fun handleValidationException(ex: MethodArgumentNotValidException): ResponseEntity<ApiResponse<Unit>> {
        // 유효성 검증 실패 항목들을 순회하며 메시지 생성
        val errors = ex.bindingResult.allErrors.map {
            ErrorResponse(
                field = (it as FieldError).field,
                value = "",
                reason = it.defaultMessage ?: "Invalid value"
            )
        }.toMutableList()

        val response = ApiResponse<Unit>(
                status = HttpStatus.BAD_REQUEST,
                message = "",
                errors = errors,
        )

        return ResponseEntity(response, HttpStatus.BAD_REQUEST)
    }
}